tests/test_datetime_generator.py

```diff
@@ -0,0 +1,53 @@
+
+
+import unittest
+from unittest import mock
+from numpy.random import Generator, PCG64
+import pandas as pd
+import numpy as np
+import synthetic_data.dataset_generators.datetime_generator as dtg
+
+
+class TestDatetimeFunctions(unittest.TestCase):
+    def setUp(self):
+        self.rng = Generator(PCG64(12345))
+        self.start_date = pd.Timestamp(2001, 12, 22)
+        self.end_date = pd.Timestamp(2023, 1, 1)
+        self.date_format_list = ["%Y-%m-%d", "%d-%m-%Y"]
+
+    def test_generate_datetime_return_type(self):
+        date_str = dtg.generate_datetime(self.rng, 
+                                        self.date_format_list[0],
+                                        self.start_date,
+                                        self.end_date)
+        self.assertIsInstance(date_str, str)
+
+    def test_generate_datetime_format(self):
+        date_str = dtg.generate_datetime(self.rng, 
+                                        self.date_format_list[0],
+                                        self.start_date,
+                                        self.end_date)
+        try:
+            pd.to_datetime(date_str, format=self.date_format_list[0])
+        except ValueError:
+            self.fail("pd.to_datetime() raised ValueError unexpectedly")
+
+    def test_generate_datetime_range(self):
+        date_str = dtg.generate_datetime(self.rng, 
+                                        self.date_format_list[0],
+                                        self.start_date,
+                                        self.end_date)
+        date_obj = pd.to_datetime(date_str, format = self.date_format_list[0])
+        self.assertTrue(self.start_date <= date_obj)
+        self.assertTrue(date_obj <= self.end_date)
+
+    @mock.patch("synthetic_data.dataset_generators.datetime_generator.generate_datetime")
+    def test_random_datetimes_return_type_and_size(self, mock_generate_datetime):
+        mock_generate_datetime.return_value = "Mocked"
```

Comment:
if we want to mock out `generate_datetime` bc of the randomness, I suggest asserting what it is called with. My main concern rn is we aren't really validating it will select from a list. 

A different way to make it deterministic, set `self.rng = mock.Mock(spec=....)` and then set the return value for the usages. We assume the np random functions work as expected.

e.g.
```python
mocked_rng = ...
mocked_rng.random.return_value = 0.5

# or use a side_effect + generator so it can change values each use
mocked_rng.random.side_effect = (x for x in [0.25, 0.5, 0.75])  # gen with 3 uses
```

Now we know what `t` will always be, we know the start \ end times, and we can do the same for formats with `choice` so we can ensure given the right inputs we know the expected output. Definitely a bit more work and maybe overkill.